modules/dnn/src/onnx/onnx_graph_simplifier.cpp

```diff
@@ -382,33 +382,63 @@ class ResizeSubgraph2 : public ExtractScalesSubgraph
     }
 };
 
-class BatchNormalizationSubgraph : public Subgraph
+class BatchNormalizationSubgraphBase : public Subgraph
 {
 public:
-    BatchNormalizationSubgraph()
+    BatchNormalizationSubgraphBase()
     {
-        int input = addNodeToMatch("");
-        int data1 = addNodeToMatch("Constant");
-        int data2 = addNodeToMatch("Constant");
-        int data3 = addNodeToMatch("Constant");
-        int data4 = addNodeToMatch("Constant");
-        int shape1 = addNodeToMatch("Constant");
-        int reshape1 = addNodeToMatch("Reshape", data1, shape1);
-        int shape2 = addNodeToMatch("Constant");
-        int reshape2 = addNodeToMatch("Reshape", data2, shape2);
+        input  = addNodeToMatch("");
+        var    = addNodeToMatch("Constant");
+        mean   = addNodeToMatch("Constant");
+        weight = addNodeToMatch("Constant");
+        bias   = addNodeToMatch("Constant");
+        A = addNodeToMatch("Constant");
+        shape1 = addNodeToMatch("Constant");
+        shape2 = addNodeToMatch("Constant");
+    }
+protected:
+    int input, var, mean, weight, bias, A, shape1, shape2;
+};
+
+class BatchNormalizationSubgraph1 : public BatchNormalizationSubgraphBase
+{
+public:
+    BatchNormalizationSubgraph1()
+    {
+        int reshape1 = addNodeToMatch("Reshape", weight, shape1);
+        int reshape2 = addNodeToMatch("Reshape", bias, shape2);
         int shape3 = addNodeToMatch("Constant");
-        int reshape3 = addNodeToMatch("Reshape", data3, shape3);
+        int reshape3 = addNodeToMatch("Reshape", var, shape3);
```

Comment:
I think it's simpler to have a boolean variable which indicates if there are Reshape over contant nodes or not

Reply:
In `BatchNormalizationSubgraph1` reshape used only for constant nodes, but in `BatchNormalizationSubgraph2` reshape used after `Mul` and `Sub` operations.